### DISCUSSION TITLE
Clarification on opCode type in event OpCodeDispatched

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#2039] [Q&A]
In the event `OpCodeDispatched` at https://github.com/nasa/fprime/blob/4f65e1ef3fa3aa44e132e6d9e7b885e840162630/Svc/CmdDispatcher/CmdDispatcher.fpp#L105 the `opCode` is `U32`, instead in any command, the opCode is `FwOpcodeType`. This is ok when `FwOpcodeType` is `U32` but from my understanding `FwOpcodeType` is project related and it could change. 

Should `FwOpcodeType` should be used also in the event  `OpCodeDispatched`? If no, why? I'm missing something?

**Comment by SterlingPeet** [comment]
Probably so.  Does your project need ``FwOpcodeType`` to be something other than U32?

I have tread the path of changing lots of abstracted types in F Prime before for my projects.  When I find specific issues that cause obvious or noticeable problems, I have submitted patches and the project has gladly introduced fixes.  That said, I haven't done this with a project that used FPP, and that involved a lot of refactoring.  Until I have a project that both uses FPP, and affords me the time to write some CI tooling that F Prime team can leverage to catch regressions, there will likely remain a fair amount of invisible bugs like this, where the typical usage resolves to the same underlying type as would be expected in other parts of the code base.

**Comment by LeStarch** [comment]
@bocchino, thoughts here?

**Comment by bocchino** [comment]
The underlying issue is that F Prime lacks the capability to specify type aliases like `FwOpcodeType = U32` in the FSW model. The definition of `FwOpcodeType` is known only to the C++ implementation. It has always been this way, even in the XML modeling days.

With FPP modeling, we can fix this issue, and it is on our agenda to fix. See https://github.com/fprime-community/fpp/issues/113.

Until we implement this fix we have to do the best we can. In the specific case of `FwOpcodeType`, opcodes are unsigned and will fit in a `U32` unless a project is using 64-bit opcodes, which seems unlikely in practice.

**Reply by SterlingPeet** [reply]
heh, and those of use that want to use 8-bit opcodes?  I guess we have to go edit some files in the code base and tweak the framer/deframer...

**Reply by bocchino** [reply]
U32 for the event arg type should work fine with 8-bit opcodes. It just wastes 3 bytes.

The framer/deframer is a separate issue. The framer/deframer code uses the C++ macro definitions, so it should accommodate changes to those definitions. However, the configurability is not well-tested. Also, until we add type aliases to FPP you do have to manually keep the C++ FSW configuration and the Python GDS configuration consistent.
